File: stack/go/nesting_score.go
package stack

// NestingScore computes score of bracket string per rules.
func NestingScore(brackets string) int {
	stack := []int{0}
	for _, ch := range brackets {
		if ch == '[' {
			stack = append(stack, 0)
		} else {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if top == 0 {
				stack[len(stack)-1]++
			} else {
				stack[len(stack)-1] += 2 * top
			}
		}
	}
	return stack[0]
}

File: stack/php/nestingScore.php
<?php

declare(strict_types=1);

function nestingScore(string $brackets): int
{
    $stack = [0];
    foreach (str_split($brackets) as $ch) {
        if ($ch === '[') {
            $stack[] = 0;
        } else {
            $top = array_pop($stack);
            if ($top === 0) {
                $stack[array_key_last($stack)] += 1;
            } else {
                $stack[array_key_last($stack)] += 2 * $top;
            }
        }
    }

    return $stack[0];
}
